frontend/mainwindow.cpp

```diff
@@ -138,3 +140,130 @@ void MainWindow::closeEvent(QCloseEvent *event)
 
     event->accept();
 }
+
+QList<QPair<QString, QString>> MainWindow::getInstruments()
+{
+    QFile file("../extra/instrumentData.xml");
+    file.open(QIODevice::ReadOnly);
+    QDomDocument dom;
+    dom.setContent(&file);
+    file.close();
+    auto rootelem = dom.documentElement();
+    auto nodelist = rootelem.elementsByTagName("inst");
+
+    QList<QPair<QString, QString>> instruments;
+    QPair<QString, QString> instrument;
+    QDomNode node;
+    QDomElement elem;
+    for (auto i = 0; i < nodelist.count(); i++)
+    {
+        node = nodelist.item(i);
+        elem = node.toElement();
+        instrument.first = elem.attribute("name");
+        instrument.second = elem.elementsByTagName("type").item(0).toElement().text();
+        instruments.append(instrument);
+    }
+    return instruments;
+}
+
+QList<QString> MainWindow::getFields(QString instrument, QString instType)
+{
+    QList<QString> bean;
+    QDomNodeList desiredInstrumentFields;
+    auto rootelem = dom_.documentElement();
+    auto instList = rootelem.elementsByTagName("inst");
+    for (auto i = 0; i < instList.count(); i++)
+    {
+        if (instList.item(i).toElement().attribute("name") == instrument)
+        {
+            desiredInstrumentFields = instList.item(i).toElement().elementsByTagName("prefFields");
+            break;
+        }
+    }
+    if (desiredInstrumentFields.isEmpty())
+    {
+        auto configDefault = rootelem.elementsByTagName(instType).item(0).toElement();
+        auto configDefaultFields = configDefault.elementsByTagName("prefFields");
+        if (configDefaultFields.isEmpty())
+        {
+            QFile file("../extra/instrumentData.xml");
```

Comment:
This gets loaded in every time the function runs.  Is that really necessary?  It feels like this should really be read and parsed on startup and then only reloaded if absolutely necessary.